Fix "reparent to new node" not remembering index

RedMser opened the pull request:
Supersedes #42730
Fixes https://github.com/godotengine/godot-proposals/issues/1652, fixes https://github.com/godotengine/godot-proposals/issues/5427

https://github.com/godotengine/godot/assets/5117197/e2526753-5571-4d63-9056-9b5f620f0fb5

Also merges the operation into a single undo step.

KoBeWi commented:
I did not debug too deeply, but the first error likely comes from here:
https://github.com/godotengine/godot/blob/c495eb5102278a110c14bbffbf833ed436d1594d/editor/scene_tree_dock.cpp#L2318
Since the action is nested, the commit will take no effect and the node is still outside the scene tree. Maybe adding an `is_inside_tree()` condition would be sufficient.

The other error comes from
https://github.com/godotengine/godot/blob/c495eb5102278a110c14bbffbf833ed436d1594d/editor/scene_tree_dock.cpp#L2415
Also caused by the node not being inside scene tree.

YuriSizov commented:
@RedMser Will you be able to look into what's discussed above?

RedMser commented:
Rebased and included the suggested fixes in a second commit, so it's easier to review.

Reparenting a MeshInstance3D causes a lot of these errors: `Node not found: "../../root/@EditorNode@17140/@Panel@13/@VBoxContainer@14/@HSplitContainer@17/@HSplitContainer@25/@HSplitContainer@33/@VBoxContainer@34/@VSplitContainer@36/@VSplitContainer@62/@VBoxContainer@63/@PanelContainer@110/MainScreen/@CanvasItemEditor@9462/@VSplitContainer@9281/@HSplitContainer@9283/@HSplitContainer@9285/@Control@9286/@SubViewportContainer@9287/@SubViewport@9288/Node/MeshInstance3D" (relative to ...`

which comes from a `_notification` call of type `NOTIFICATION_ENTER_TREE`. So something seems pretty wrong here still, and these is_inside_tree checks I added are likely not a proper fix.

Sorry I don't have much time to work on this lately, but I hope it can be resolved and fixed for 4.3

KoBeWi commented:
The errors are caused due to updating NodePath properties. It seems to happen with other nodes too, e.g. I tried reparenting AnimationPlayer and ended with this:
![image](https://github.com/godotengine/godot/assets/2223172/a8c481a5-f01c-4206-9af3-8f64746b9f3d)
Not sure why is that.

RedMser committed:
Fix "reparent to new node" not remembering index

RedMser commented:
Well, I got the PR into a better state than it previously was.

I moved any logic that relied on the node already being added to the scene tree, and put it into `add_do_method "_post_do_create"`

No errors anymore from my testing, and NodePath in AnimationPlayer is properly updated, however this did break "Reparent to New Node" no longer working in live reloading context. I wrote a TODO comment, best solution I can think of is custom do/undo methods so that the paths are evaluated only once child was added, but this all feels really hacky and not worth it...

KoBeWi reviewed (approved):
It works correctly now.
But as you noted, it breaks live edit for Reparent to New Node, while it was functional before. I think it's less important than the linked issues though.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -2354,8 +2353,12 @@ void SceneTreeDock::_do_reparent(Node *p_new_parent, int p_position_in_parent, V
 
-		undo_redo->add_do_method(ed, "live_debug_reparent_node", edited_scene->get_path_to(node), edited_scene->get_path_to(new_parent), new_name, new_position_in_parent);
-		undo_redo->add_undo_method(ed, "live_debug_reparent_node", NodePath(String(edited_scene->get_path_to(new_parent)).path_join(new_name)), edited_scene->get_path_to(node->get_parent()), node->get_name(), node->get_index(false));
+		// TODO: Live reloading of "Reparent to New Node" is broken.
```
```suggestion
		// TODO: Live editing for "Reparent to New Node" option is broken.
```

akien-mga commented:
Thanks!

kleonc commented on editor/scene_tree_dock.cpp:
```diff
@@ -2922,22 +2929,24 @@ void SceneTreeDock::_create() {
+		int original_position = -1;
 		if (only_one_top_node) {
 			parent = top_node->get_parent();
+			original_position = top_node->get_index();
```
@RedMser This should be `original_position = top_node->get_index(false);`, as `move_child` takes in an index not accounting for internal nodes.

Here's an example of it failing for TabContainer (which has front-internal TabBar child):
![godot windows editor x86_64_IqOzGwV6GD](https://github.com/user-attachments/assets/e6f76b41-863a-4bb7-ba3b-aca954af46a3)
After `First` -> `Reparent to New Node...`:
![godot windows editor x86_64_NDmkEd7XBO](https://github.com/user-attachments/assets/6b77b116-e132-461b-9148-e4dbd94f6c07)
Correct result (currently only if parent has no internal children):
![godot windows editor x86_64_z8THw1XRG6](https://github.com/user-attachments/assets/1bfdcfc4-876b-49d4-91a7-6eae5a7f3cef)

RedMser replied:
Ah I see. Would you like to create a followup PR to fix this, or should I do so?

kleonc replied:
Go ahead! :slightly_smiling_face:

RedMser replied:
Done, thanks for the info!
https://github.com/godotengine/godot/pull/96646

kleonc commented:
Note: should be cherry-picked together with #96646.

akien-mga commented:
Cherry-picked for 4.3.1.
